cc/position.cc

```diff
@@ -424,6 +429,264 @@ float Position::CalculateScore(float komi) {
+// alive black regions as follows:
+//  - Let X be the set of all black chains.
+//  - Let R be the set of small black-enclosed regions of X.
+//  - Iterate the following two steps until neither one removes an item:
+//    - Remove from X all black chains with fewer than two vital
+//      black-enclosed regions in R.
+//    - Remove from R all black-enclosed regions with a surrounding stone in a
+//      chain not in X.
+//
+// Unconditionally alive chains are also called pass-alive because they cannot
+// be captured by the opponent even if that player always passes on their turn.
+// More details:
+//   https://senseis.xmp.net/?BensonsDefinitionOfUnconditionalLife
+std::array<Color, kN * kN> Position::CalculatePassAliveRegions(
+    Color color) const {
+  struct BensonGroup {
+    Coord first_stone = Coord::kInvalid;
+    std::set<Coord> liberties;
+    int num_vital_regions = 0;
+    bool is_pass_alive = false;
+  };
+
+  struct BensonRegion {
+    Coord first_stone = Coord::kInvalid;
+    std::set<Coord> empty_points;
+    std::set<Coord> other_color_points;
+    std::set<BensonGroup*> vital_for;
+    std::set<BensonGroup*> enclosing_groups;
+    bool is_pass_alive = false;
+  };
+
+  std::vector<std::unique_ptr<BensonRegion>> region_pool;
+  std::vector<std::unique_ptr<BensonGroup>> group_pool;
+  std::array<BensonRegion*, kN * kN> regions;
+  std::array<BensonGroup*, kN * kN> groups;
+
+  // Build the set of all groups.
+  board_visitor_->Begin();
+  for (int row = 0; row < kN; ++row) {
+    for (int col = 0; col < kN; ++col) {
+      Coord c(row, col);
+      if (stones_[c].color() != color || !board_visitor_->Visit(c)) {
+        continue;
+      }
+
+      auto g = absl::make_unique<BensonGroup>();
+      g->first_stone = c;
+      while (!board_visitor_->Done()) {
+        c = board_visitor_->Next();
+        groups[c] = g.get();
+
+        for (auto nc : kNeighborCoords[c]) {
+          auto ns = stones_[nc];
+          if (ns.empty()) {
+            g->liberties.insert(nc);
+          } else if (ns.color() == color) {
+            board_visitor_->Visit(nc);
+          }
+        }
+      }
+      group_pool.push_back(std::move(g));
+    }
+  }
+
+  // Build the set of all regions.
+  board_visitor_->Begin();
+  for (int row = 0; row < kN; ++row) {
+    for (int col = 0; col < kN; ++col) {
+      Coord c(row, col);
+      if (stones_[c].color() == color || !board_visitor_->Visit(c)) {
+        continue;
+      }
+
+      auto r = absl::make_unique<BensonRegion>();
+      r->first_stone = c;
+      while (!board_visitor_->Done()) {
+        c = board_visitor_->Next();
+
+        regions[c] = r.get();
+        if (stones_[c].empty()) {
+          r->empty_points.insert(c);
+        } else {
+          r->other_color_points.insert(c);
+        }
+
+        for (auto nc : kNeighborCoords[c]) {
+          auto ns = stones_[nc];
+          if (ns.color() != color) {
```

Comment:
random question: why is .color() a computed property rather than just dereferencing a struct field? Is it doing a lookup in the parent board object or something like that?

Reply:
A `Stone` is a bit-packed `uint16` that contains the colour in the bottom 2 bits and an ID of the group that the stone belongs to in the upper 14 bits.